main.go

```diff
@@ -0,0 +1,517 @@
+		Package:  p,
+		Indent:   c.Indent + 1,
+		Writer:   c.Writer,
+	}
+}
+
+func (c *Context) WriteLn(line string) {
+	for i := 0; i < c.Indent; i++ {
+		c.Writer.Write([]byte{' ', ' '})
+	}
+	c.Writer.Write([]byte(line))
+	c.Writer.Write([]byte{'\n'})
+}
+
+func Field_GetType(p *Package, f *ast.Field) types.Type {
+	var unwrap func(t ast.Expr) *ast.Ident
+
+	// `p.TypesInfo.Defs` maps `*ast.Ident` to `types.Type`. Sometimes `f.type`
+	// is encapsulated by an "intermediate" type such as "star expression" or
+	// "selector expression". Fortunately, these can be unwrapped to reveal
+	// the underlying `*ast.Ident`.
+	unwrap = func(t ast.Expr) *ast.Ident {
+		switch x := t.(type) {
+		case *ast.SelectorExpr:
+			return unwrap(x.Sel)
+		case *ast.StarExpr:
+			return unwrap(x.X)
+		case *ast.Ident:
+			return x
+		}
+		return nil
+	}
+
+	if len(f.Names) > 0 {
+		return p.TypesInfo.Defs[f.Names[0]].Type()
+	} else {
+		return p.TypesInfo.Defs[unwrap(f.Type)].Type()
+	}
+}
+
+func Type_IsArray(t types.Type) bool {
+	// In most circumstances the "intermediate" types matter. However,
+	// `toml` automatically constructs the intermediates when parsing.
+	// For instance, if a struct contains a field of type `*string`,
+	// `toml` can initialize it from a plain-old "string". We will
+	// ignore the intermediates when appropriate.
+	switch x := t.(type) {
+	case *types.Pointer:
+		return Type_IsArray(x.Elem())
+	case *types.Slice:
+		return true
+	}
+	return false
+}
+
+func Type_IsStruct(t types.Type) bool {
+	x := Type_GetBasal(t)
+
+	if _, ok := x.(*types.Struct); ok {
+		return true
+	}
+	if _, ok := x.Underlying().(*types.Struct); ok {
+		return true
+	}
+	return false
+}
+
+func Type_GetBasal(t types.Type) types.Type {
+	// In most circumstances the "intermediate" types matter. However,
+	// `toml` automatically constructs the intermediates when parsing.
+	// For instance, if a struct contains a field of type `*string`,
+	// `toml` can initialize it from a plain-old "string". We will
+	// ignore the intermediates when appropriate.
+	switch x := t.(type) {
+	case *types.Pointer:
+		return Type_GetBasal(x.Elem())
+	case *types.Slice:
+		return Type_GetBasal(x.Elem())
+	}
+	return t
+}
+
+func Type_GetNamed(t types.Type) *types.Named {
+	x, ok := Type_GetBasal(t).(*types.Named)
+	if !ok {
+		return nil
+	}
+	return x
+}
+
+func Type_LoadPackage(pkgs *map[string]Package, t *types.Named) *Package {
+	path := t.Obj().Pkg().Path()
+
+	if p, ok := (*pkgs)[path]; ok {
+		return &p
+	}
+
+	config := loader.Config{
+		ParserMode: parser.ParseComments,
+	}
+	config.Import(path)
+	prog, err := config.Load()
+	if err != nil {
+		panic(err)
+	}
+
+	for _, pkgInfo := range prog.InitialPackages() {
+		(*pkgs)[path] = Package{
+			Fset:      prog.Fset,
+			Files:     pkgInfo.Files,
+			Path:      pkgInfo.Pkg.Path(),
+			TypesInfo: &pkgInfo.Info,
+		}
+	}
+
+	if p, ok := (*pkgs)[path]; ok {
+		return &p
+	}
+
+	return nil
+}
+
+func WriteIndent(context *Context) {
```

Comment:
Why is the indent in WriteLn two spaces and this one a single space?

Reply:
Should this be used in `WriteLn`?

Reply:
I think that this may be how GitHub is rendering the text. When I open it in my editor, `WriteIndent` is emitting two spaces. However, I was using `WriteIndent`, before I had written `WriteLn` and I had forgotten to remove it once I switched everything over to `WriteLn`. I've pushed a commit to remove it.